doc: Add ACRN secureboot setup with GRUB (#5532)
projectacrn/acrn-hypervisor: doc/tutorials/acrn-secure-boot-with-grub.rst

.. _how-to-enable-acrn-secure-boot-with-grub:

Enable ACRN Secure Boot with GRUB
#################################

This document shows how to enable ACRN secure boot with GRUB including:

  -  ACRN Secure Boot Sequence
  -  Generate GPG Key
  -  Setup Standalone GRUB EFI Binary
  -  UEFI Secure Boot

  Validation Environment:

> 	Hardware Platform: INTEL WHL

Reviewer: Can you change this to a pointer to the platform from the "Supported Hardware" list?

If we expect that this method can work with other platforms too, it would be good to add a note in that sense.

	ACRN Scenario: Industry
	Service VM: Ubuntu 18.04

> 	GRUB: 2.02

Reviewer: We use 2.04 with a patch in our main Getting Started Guide... can we use that one too here? I would prefer that we stick to a single version if at all possible.

Author: @gvancuts  i tested using 2.02,  QA team will re-try these steps with 2.04, i can update it after that.  make sense?

Reviewer: > @gvancuts i tested using 2.02, QA team will re-try these steps with 2.04, i can update it after that. make sense?

Do you know when they will test this? I would like them to test the version 2.04 **with** the patch we use in our GSG too. Can they do that?

Author: we have tested with grub 2.04, thanks

> Note: Please make sure you have alternative ways to restore GRUB when you got stuck.

Reviewer: I would rephrase this into something like: "Note that Grub may stop booting in case of problems, make sure you know how to recover a bootloader on your platform". @dbkinder , any better suggestion / (Not quite happy with what I wrote)

Reviewer: Use `.. note::` here for proper rendering of the note.

ACRN Secure Boot Sequence
*************************

  ACRN can be booted by Multiboot compatible bootloader, following diagram illustrates the boot sequence of ACRN with GRUB:

>      .. image:: images/acrn-secureboot-flow.png

Reviewer: Can you update the diagram to use "Service VM kernel" instead of "SOS kernel"?

        :align: center
        :width: 600px

  Refer to :ref:`using_grub` for details to enable GRUB on ACRN.

  From point view of secureboot:
	- UEFI firmware verifies Shim/GRUB

> 	- GRUB verifies ACRN, Service VM kernel and Prelaunch VM kernel

Reviewer: ```suggestion
	- GRUB verifies ACRN, Service VM kernel and pre-launched User VM kernel
```

> 	- SOS kernel verifies DM and UOS OVMF thru ACRN DM

Reviewer: ```suggestion
	- Service VM OS kernel verifies the Device Model (`acrn-dm`) and User VM OVMF bootloader (with the help of `acrn-dm`)
```

Would this still be correct?

> 	- UOS OVMF starts the guest side verified boot process

Reviewer: ```suggestion
	- User VM OVMF starts the guest side verified boot process
```

I'm wondering if it wouldn't be better to talk about the "virtual bootloader" instead of "User VM OVMF". @dbkinder , thoughts?

>   This document provides guide to enable verification function of GRUB to verify ACRN binaries(such acrn.bin, service VM bzImage and Prelaunch VM kernel image if it is present).

Reviewer: ```suggestion
  This document provides guide to enable verification function of GRUB to verify ACRN binaries (such `acrn.bin`, Service VM kernel (`bzImage`) and pre-launched User VM kernel image if it is present).
```

Generate GPG Key
****************

  GRUB supports loading of GPG signed files only if digital signatures enabled.
  Example to generate GPG signing key::

>   $mkdir --mode 0700 keys

Reviewer: ```suggestion
  $ mkdir --mode 0700 keys
```
Do this too for the lines below.

I suspect this code-block formatting is not rendered as expected, have you verified the final rendering (using `make doc`)?

Author: @gvancuts my local make doc does not work,  not see error reported from CI /build doc, are they same?

//my local build doc err, any hint?

make[1]: Entering directory '/home/yh/workspace/acrn/acrn-hypervisor/doc'
Makefile:60: recipe for target 'html' failed
make[1]: [html] Error 2 (ignored)

New errors/warnings found, please fix them:
==============================================

Extension error:
Could not import extension sphinx.builders.applehelp (exception: cannot import name 'RemovedInSphinx40Warning')
Makefile:60: recipe for target 'html' failed
make[1]: *** [html] Error 1
make[1]: Leaving directory '/home/yh/workspace/acrn/acrn-hypervisor/doc'
Makefile:184: recipe for target 'doc' failed
make: *** [doc] Error 2

  $gpg --homedir keys  --gen-key
  $gpg --homedir keys --export > boot.key

  ``gpg --gen-key`` generates key pair, private key will be used to sign GRUB configuration files and ACRN binaries, public key will be embedded in GRUB and is used to verify GRUB configuration files or binaries GRUB trys to load.

Setup Standalone GRUB EFI Binary
********************************

Prepare Initial GRUB Configuration grub.init.cfg
================================================

>   Store the following minimal grub config in grub.init.cfg, ``set check_signatures =enforce`` is to enable digital signatures in GRUB::

Reviewer: ```suggestion
  Store the following minimal grub config in grub.init.cfg, ``set check_signatures=enforce`` is to enable digital signatures in GRUB::
```

>   Store the following minimal grub config in grub.init.cfg, ``set check_signatures =enforce`` is to enable digital signatures in GRUB::

Reviewer: I think @dbkinder prefers that we use :file:`grub.init.cfg` for files.

Where is that file located by the way?

Author: @gvancuts  it is a tmp file and will be built into grub finally.

	set check_signatures=enforce
	export check_signatures

	search --no-floppy --fs-uuid --set=root ESP_UUID
	configfile /grub.cfg
	echo /grub.cfg did not boot the system, rebooting in 10 seconds.
	sleep 10
	reboot

>   Replace the ESP_UUID with UUID of your EFI system partition, with the following output of ``lsblk -f``, replace it with ``24FC-BE7A``.::

Reviewer: @dbkinder can you help reword this sentence?

Also, just wondered where we should replace the double-backticks around `lbblk` by :command:`lsblk -f`?

  the ``--pubkey`` option adds a GPG public key which will be used for verification, and public key ``boot.key``
  is not required anymore after ``grub-mkstandalone``, hence it can be removed.

  Script Example::

	#!/bin/bash
	#

	TARGET_EFI='/boot/efi/EFI/ubuntu/grubx64.efi'

	# GRUB doesn't allow loading new modules from disk when secure boot is in
	# effect, therefore pre-load the required modules.
	MODULES="all_video archelp boot bufio configfile crypto echo efi_gop efi_uga ext2 extcmd  \
	fat font fshelp gcry_dsa gcry_rsa gcry_sha1 gcry_sha512 gettext gfxterm linux linuxefi ls \
	memdisk minicmd mmap mpi normal part_gpt part_msdos password_pbkdf2 pbkdf2 reboot relocator \
	search search_fs_file search_fs_uuid search_label sleep tar terminal verify video_fb"

	grub-mkstandalone \
		--directory /usr/lib/grub/x86_64-efi \
		--format x86_64-efi \
		--modules "$MODULES" \
		--pubkey ./boot.key \
		--output ./grubx64.efi  \
		"boot/grub/grub.cfg=./grub.init.cfg" \
		"boot/grub/grub.cfg.sig=./grub.init.cfg.sig"

	echo "writing signed grub.efi to '$TARGET_EFI'"
	sudo cp ./grubx64.efi "$TARGET_EFI"

>   Note: You should backup current GRUB image in '$TARGET_EFI', then you can restore your grub in case you got stuck after updating it.

Reviewer: Replace with `.. note::`

Prepare grub.cfg
================
  Define the menuenty for your system in a new GRUB configration file ``grub.cfg``.
  Example::

	# @/boot/efi/grub.cfg for grub secure boot
	 set timeout_style=menu
	 set timeout=5
	 set gfxmode=auto
	 set gfxpayload=keep
	 terminal_output gfxterm

	 menuentry "ACRN Multiboot Ubuntu Service VM" --users "" --id ubuntu-service-vm {

	 search --no-floppy --fs-uuid --set 3df12ea1-ef12-426b-be98-774665c7483a

	 echo 'loading ACRN...'
	 multiboot2 /boot/acrn/acrn.bin  root=PARTUUID="c8ee7d92-8935-4e86-9e12-05dbeb412ad6"
	 module2 /boot/bzImage Linux_bzImage
	}

  Update the UUID (``--set``) and PARTUUID (``root=`` parameter) of the root partition (e.g. /dev/nvme0n1p2) according to
  your your hardware(e.g. use ``blkid`` tool).

>   Copy grub.cfg to root of ESP(e.g. /boot/efi/EFI/).

Reviewer: ```suggestion
  Copy :file:`grub.cfg` to your ESP (e.g. `/boot/efi/EFI/`).
```

Sign grub.cfg and ACRN Binaries
===============================

>   grub.cfg file and all ACRN binaries that will be loaded by GRUB MUST be signed with GPG key.

Reviewer: ```suggestion
  The :file:`grub.cfg` file and all ACRN binaries that will be loaded by GRUB **must** be signed with GPG key.
```

  Sign script example::

	gpg --homedir keys --detach-sign path/to/grub.cfg
	gpg --homedir keys --detach-sign path/to/acrn.bin
	gpg --homedir keys --detach-sign path/to/sos_kernel/bzImage

  Now, you can reboot and system will boot with single GRUB EFI binary.
  GRUB will refuse to boot if any files to be loaded are tampered.

UEFI Secure Boot
================

  UEFI secure boot can be enabled to verify GRUB EFI binary, which may be verified by SHIM or UEFI firmware (without SHIM).
  Refer to `How to sign things for Secure Boot <https://ubuntu.com/blog/how-to-sign-things-for-secure-boot>`_

>   for detailed steps(not verified).

Reviewer: ```suggestion
  for detailed steps (not verified).
```

.. _how-to-enable-acrn-secure-boot-with-grub:

Enable ACRN Secure Boot with GRUB
#################################

This document shows how to enable ACRN secure boot with GRUB including:

  -  ACRN Secure Boot Sequence
  -  Generate GPG Key
  -  Setup Standalone GRUB EFI Binary
  -  UEFI Secure Boot

  Validation Environment:
	Hardware Platform: WHL-IPC-I7, Supported hardware described in :ref:`hardware`.
	ACRN Scenario: Industry
	Service VM: Ubuntu 18.04
	GRUB: 2.02

> ..note::

Reviewer: The formatting seems off, it should be something like:
```
.. note::
   Grub may stop... blah blah
```

	Note that GRUB may stop booting in case of problems, make sure you know how to recover a bootloader on your platform.

ACRN Secure Boot Sequence
*************************

  ACRN can be booted by Multiboot compatible bootloader, following diagram illustrates the boot sequence of ACRN with GRUB:

     .. image:: images/acrn_secureboot_flow.png
        :align: center
        :width: 600px

  Refer to :ref:`using_grub` for details to enable GRUB on ACRN.

  From point view of secureboot:
	- UEFI firmware verifies Shim/GRUB
	- GRUB verifies ACRN, Service VM kernel and pre-launched User VM kernel

> 	- Service VM OS kernel verifies the Device Model (`acrn-dm`) and User VM OVMF bootloader (with the help of `acrn-dm`)

Reviewer: ```suggestion
	- Service VM OS kernel verifies the Device Model (``acrn-dm``) and User VM OVMF bootloader (with the help of ``acrn-dm``)
```

	- User VM virtual bootloader (e.g. OVMF) starts the guest side verified boot process

>    This document provides guide to enable verification function of GRUB to verify ACRN binaries (such `acrn.bin`, Service VM kernel (`bzImage`) and pre-launched User VM kernel image if it is present).

Reviewer: ```suggestion
   This document provides guide to enable verification function of GRUB to verify ACRN binaries (such ``acrn.bin``, Service VM kernel (``bzImage``) and pre-launched User VM kernel image if it is present).
```

Generate GPG Key
****************

  GRUB supports loading of GPG signed files only if digital signatures enabled.

>   Example to generate GPG signing key::

Reviewer: I _think_ you need a blank line here just before the code-block. Do you run `make doc` after your updates?

  $ mkdir --mode 0700 keys
  $ gpg --homedir keys  --gen-key
  $ gpg --homedir keys --export > boot.key

  ``gpg --gen-key`` generates key pair, private key will be used to sign GRUB configuration files and ACRN binaries, public key will be embedded in GRUB and is used to verify GRUB configuration files or binaries GRUB trys to load.

Setup Standalone GRUB EFI Binary
********************************

Prepare Initial GRUB Configuration grub.init.cfg
================================================

>   Create file ``grub.init.cfg`` to store the following minimal grub configuration, ``set check_signatures=enforce`` is to enable digital signatures in GRUB::

Reviewer: @dbkinder can provide more details but ReST is sensitive to indentation, I'm wondering what it does when it sees that sentence starting with three spaces.

  the ``--pubkey`` option adds a GPG public key which will be used for verification, and public key ``boot.key``
  is not required anymore after ``grub-mkstandalone``, hence it can be removed.

  Script Example::

	#!/bin/bash
	#

	TARGET_EFI='/boot/efi/EFI/ubuntu/grubx64.efi'

	# GRUB doesn't allow loading new modules from disk when secure boot is in
	# effect, therefore pre-load the required modules.
	MODULES="all_video archelp boot bufio configfile crypto echo efi_gop efi_uga ext2 extcmd  \
	fat font fshelp gcry_dsa gcry_rsa gcry_sha1 gcry_sha512 gettext gfxterm linux linuxefi ls \
	memdisk minicmd mmap mpi normal part_gpt part_msdos password_pbkdf2 pbkdf2 reboot relocator \
	search search_fs_file search_fs_uuid search_label sleep tar terminal verify video_fb"

	grub-mkstandalone \
		--directory /usr/lib/grub/x86_64-efi \
		--format x86_64-efi \
		--modules "$MODULES" \
		--pubkey ./boot.key \
		--output ./grubx64.efi  \
		"boot/grub/grub.cfg=./grub.init.cfg" \
		"boot/grub/grub.cfg.sig=./grub.init.cfg.sig"

	echo "writing signed grub.efi to '$TARGET_EFI'"
	sudo cp ./grubx64.efi "$TARGET_EFI"

> ..note::

Reviewer: Same here, indentation is incorrect.

	You should backup current GRUB image in '$TARGET_EFI', then you can restore your grub in case you got stuck after updating it.

Prepare grub.cfg
================
  Define the menuenty for your system in a new GRUB configration file ``grub.cfg``.
  Example::

	# @/boot/efi/grub.cfg for grub secure boot
	 set timeout_style=menu
	 set timeout=5
	 set gfxmode=auto
	 set gfxpayload=keep
	 terminal_output gfxterm

	 menuentry "ACRN Multiboot Ubuntu Service VM" --users "" --id ubuntu-service-vm {

	 search --no-floppy --fs-uuid --set 3df12ea1-ef12-426b-be98-774665c7483a

	 echo 'loading ACRN...'
	 multiboot2 /boot/acrn/acrn.bin  root=PARTUUID="c8ee7d92-8935-4e86-9e12-05dbeb412ad6"
	 module2 /boot/bzImage Linux_bzImage
	}

  Update the UUID (``--set``) and PARTUUID (``root=`` parameter) of the root partition (e.g. /dev/nvme0n1p2) according to
  your your hardware(e.g. use ``blkid`` tool).

  Copy :file:`grub.cfg` to your ESP (e.g. `/boot/efi/EFI/`).

Sign grub.cfg and ACRN Binaries
===============================

  The :file:`grub.cfg` file and all ACRN binaries that will be loaded by GRUB **must** be signed with GPG key.

>   Sign script example::

Reviewer: Better call that a "sign sequence example". It cannot really be copy-pasted into a file and ran as a script

.. image:: images/acrn_secureboot_flow.png
   :align: center
   :width: 800px

For details on enabling grub on ACRN, see :ref:`using_grub`.

From a secureboot point of view:

- UEFI firmware verifies shim/grub
- grub verifies ACRN, Service VM kernel, and pre-launched User VM kernel
- Service VM OS kernel verifies the Device Model (``acrn-dm``) and User
  VM OVMF bootloader (with the help of ``acrn-dm``)
- User VM virtual bootloader (e.g. OVMF) starts the guest side verified boot process

This document shows you how to enable grub to
verify ACRN binaries such ``acrn.bin``, Service VM kernel (``bzImage``), and
if present, a pre-launched User VM kernel image.

.. rst-class:: numbered-step

Generate GPG Key
****************

grub supports loading GPG signed files only if digital signatures are
enabled. Here's an example of generating a GPG signing key::

  mkdir --mode 0700 keys
  gpg --homedir keys  --gen-key
  gpg --homedir keys --export > boot.key

The ``gpg --gen-key`` command generates a public and private key pair.
The private key is used to sign grub configuration files and ACRN
binaries. The public key will be embedded in grub and is used to verify

> grub configuration files or binaries grub trys to load.

Reviewer 2: oops :)

.. _how-to-enable-acrn-secure-boot-with-grub:

Enable ACRN Secure Boot with GRUB
#################################

This document shows how to enable ACRN secure boot with GRUB including:

-  ACRN Secure Boot Sequence
-  Generate GPG Key
-  Setup Standalone GRUB EFI Binary
-  Enable UEFI Secure Boot

**Validation Environment:**

- Hardware Platform: WHL-IPC-I7, Supported hardware described in
  :ref:`hardware`.
- ACRN Scenario: Industry

> - Service VM: Yocto

Reviewer: This mentions Yocto, although I believe it mentionned Ubuntu previously.

The script below still seems to be for Ubuntu, e.g. `TARGET_EFI='/boot/efi/EFI/ubuntu/grubx64.efi'`

Author: @gvancuts , tested on both yoctor and ubuntu, have update this mismatch, thanks.

    sleep 10
    reboot

Replace the ESP_UUID with the UUID of your EFI system partition (found
by running the :command:`lsblk -f`.  In the example output below,
the UUID is ``24FC-BE7A``:

.. code-block:: console
    :emphasize-lines: 2

    sda
    ├─sda1 vfat   ESP    24FC-BE7A                            /boot/efi
    ├─sda2 vfat   OS     7015-557F
    ├─sda3 ext4   UBUNTU e8640994-b2a3-45ad-9b72-e68960fb22f0 /
    └─sda4 swap          262d1113-64be-4910-a700-670b9d2277cc [SWAP]

Enable Authentication in GRUB
=============================

With authentication enabled, a user/password is required to restrict
access to the GRUB shell, where arbitrary commands could be run.
A typical GRUB configuration fragment (added to ``grub.init.cfg``) might
look like this::

    set superusers="root"
    export superusers
    password_pbkdf2 root GRUB_PASSWORD_HASH

Replace the ``GRUB_PASSWORD_HASH`` with the result of the :command:`grub-mkpasswd-pbkdf2`
with your custom passphrase.

Use this command to sign the :file:`grub.init.cfg` file with your private
GPG key and create the :file:`grub.init.cfg.sig`::

    gpg --homedir keys --detach-sign grub.init.cfg

> Create a single GRUB EFI Binary

Reviewer 2: I'm thinking this should be ``signed`` and not ``single`` here, and on line 126 below

Author: @dbkinder, not signed, suppose "standalone" is better, right?

    This would allow you to restore GRUB in case of errors updating it.

Here's an example script to do this build::

    #!/bin/bash
    #

    TARGET_EFI='/boot/efi/EFI/ubuntu/grubx64.efi'

    # GRUB doesn't allow loading new modules from disk when secure boot is in
    # effect, therefore pre-load the required modules.

    MODULES="all_video archelp boot bufio configfile crypto echo efi_gop efi_uga ext2 extcmd  \
     fat font fshelp gcry_dsa gcry_rsa gcry_sha1 gcry_sha512 gettext gfxterm linux linuxefi ls \
     memdisk minicmd mmap mpi normal part_gpt part_msdos password_pbkdf2 pbkdf2 reboot relocator \
     search search_fs_file search_fs_uuid search_label sleep tar terminal verifiers video_fb"

    grub-mkstandalone \
     --directory /usr/lib/grub/x86_64-efi \
     --format x86_64-efi \
     --modules "$MODULES" \
     --pubkey ./boot.key \
     --output ./grubx64.efi  \
     "boot/grub/grub.cfg=./grub.init.cfg" \
     "boot/grub/grub.cfg.sig=./grub.init.cfg.sig"

    echo "writing signed grub.efi to '$TARGET_EFI'"
    sudo cp ./grubx64.efi "$TARGET_EFI"

.. rst-class:: numbered-step

Prepare grub.cfg
****************

> Define the menu entry for your system in a new GRUB configuration :file: `grub.cfg`.

Reviewer: Wouldn't that file be overwritten by someone doing a `sudo update-grub`?

Author: Will not, path/to/this/grub.cfg will not be touched by 'update-grub', unless user intend to do so, e.g. put it under /boot/grub/

Reviewer: OK, thanks!

    openssl req -new -x509 -newkey rsa:2048 -subj "/CN=KEK/" -keyout KEK.key -out KEK.crt -days 7300 -nodes -sha256
    openssl req -new -x509 -newkey rsa:2048 -subj "/CN=db/"  -keyout db.key  -out db.crt  -days 7300 -nodes -sha256

-Convert ``*.crt`` keys to the ESL format understood for UEFI::

    cert-to-efi-sig-list PK.crt PK.esl
    cert-to-efi-sig-list KEK.crt KEK.esl
    cert-to-efi-sig-list db.crt db.esl

-Sign ESL files::

    sign-efi-sig-list -k PK.key -c PK.crt PK PK.esl PK.auth
    sign-efi-sig-list -k PK.key -c PK.crt KEK KEK.esl KEK.auth
    sign-efi-sig-list -k KEK.key -c KEK.crt db db.esl db.auth

The keys to be enrolled in UEFI firmware: :file:`PK.der` , :file:`KEK.der` , :file:`db.der`.
The keys to sign bootloader image :file:'grubx64.efi': :file:`db.key` , :file:`db.crt`.

Sign GRUB Image With ``db`` Key
================================

    sbsign --key db.key --cert db.crt path/to/grubx64.efi

:file:`grubx64.efi.signed` will be created, it will be your bootloader.

Enroll UEFI Keys To UEFI Firmware
=================================

Enroll ``PK``(:file:`PK.der`), ``KEK``(:file:`KEK.der`) and ``db``(:file:`db.der`) in

> Secure Boot Configuration UI, which depends on your platform UEFI firmware. From now on,

Reviewer 2: We should include an example of how to "Enroll" these files on one of our supported platforms, for example, the one you validated listed at the top of the page.

Author: @dbkinder  have updated this part, please review, thank you.
-Yonghua

Now, you can reboot and the system will boot with the signed GRUB EFI binary.
GRUB will refuse to boot if any files it attempts to load have been tampered
with.

.. rst-class:: numbered-step

Enable UEFI Secure Boot
***********************

Creating UEFI Secure Boot Key
=============================

-Generate your own keys for Secure Boot::

    openssl req -new -x509 -newkey rsa:2048 -subj "/CN=PK/"  -keyout PK.key  -out PK.crt  -days 7300 -nodes -sha256
    openssl req -new -x509 -newkey rsa:2048 -subj "/CN=KEK/" -keyout KEK.key -out KEK.crt -days 7300 -nodes -sha256
    openssl req -new -x509 -newkey rsa:2048 -subj "/CN=db/"  -keyout db.key  -out db.crt  -days 7300 -nodes -sha256

-Convert ``*.crt`` keys to the ESL format understood for UEFI::

    cert-to-efi-sig-list PK.crt PK.esl
    cert-to-efi-sig-list KEK.crt KEK.esl
    cert-to-efi-sig-list db.crt db.esl

-Sign ESL files::

    sign-efi-sig-list -k PK.key -c PK.crt PK PK.esl PK.auth
    sign-efi-sig-list -k PK.key -c PK.crt KEK KEK.esl KEK.auth
    sign-efi-sig-list -k KEK.key -c KEK.crt db db.esl db.auth

The keys to be enrolled in UEFI firmware: :file:`PK.der` , :file:`KEK.der` , :file:`db.der`.

> The keys to sign bootloader image :file:'grubx64.efi': :file:`db.key` , :file:`db.crt`.

Reviewer: ```suggestion
The keys to sign bootloader image :file:`grubx64.efi`: :file:`db.key` , :file:`db.crt`.
```
